engine/gamesys/src/gamesys/scripts/script_buffer.cpp

```diff
@@ -879,6 +880,209 @@ namespace dmGameSystem
+                    case dmBuffer::VALUE_TYPE_INT16:
+                        values = (void*) luaTableToArray<int16_t>(L, count, valueType);
+                    break;
+                    case dmBuffer::VALUE_TYPE_INT32:
+                        values = (void*) luaTableToArray<int32_t>(L, count, valueType);
+                    break;
+                    case dmBuffer::VALUE_TYPE_INT64:
+                        values = (void*) luaTableToArray<int64_t>(L, count, valueType);
+                    break;
+                    default:
+                        return DM_LUA_ERROR("%s.%s invalid value type supplied: %ld.", SCRIPT_LIB_NAME, SCRIPT_TYPE_NAME_BUFFERMETADATA, valueTypeRaw);
+                }
+
+                dmBuffer::Result r = dmBuffer::SetMetaData(hbuffer, entry_name, (void*) values, count, valueType);
+                free(values);
+
+                if( r != dmBuffer::RESULT_OK )
+                {
+                    return DM_LUA_ERROR("%s.%s: cannot set metadata for buffer: %s", SCRIPT_LIB_NAME, SCRIPT_TYPE_NAME_BUFFERMETADATA, dmBuffer::GetResultString(r));
+                }
+            } else {
+                return DM_LUA_ERROR("%s.%s: invalid metadata", SCRIPT_LIB_NAME, SCRIPT_TYPE_NAME_BUFFERMETADATA);
+            }
+        } else
+            return DM_LUA_ERROR("%s.%s: invalid number of arguments", SCRIPT_LIB_NAME, SCRIPT_TYPE_NAME_BUFFERMETADATA);
+
+        return 0;
+    }
+
+    // copies values from array to (empty) lua table on top of the stack
+    template <typename T>
+    static void arrayToLuaTable(lua_State* L, const T* values, uint32_t count, dmBuffer::ValueType valueType) {
+        if (valueType == dmBuffer::VALUE_TYPE_FLOAT32) {
+            for (uint32_t i=0; i<count; i++) {
+                T value = values[i];
+                lua_pushnumber(L, value);
+                lua_rawseti(L, -2, i+1); // lua indices start from 1
+            }
+        } else {
+            for (uint32_t i=0; i<count; i++) {
+                T value = values[i];
+                lua_pushinteger(L, value);
+                lua_rawseti(L, -2, i+1); // lua indices start from 1
+            }
+        }
+    }
+
+    // local metadata, numbertype  = buffer.get_metadata(self.new_buffer, "aabb")
+    static int GetMetadata(lua_State* L)
+    {
+        DM_LUA_STACK_CHECK(L, 2); // return metadata table and value type
+        int top = lua_gettop(L);
+
+        if (top == 2)
+        {
+            // get buffer
+            dmScript::LuaHBuffer* buffer = dmScript::CheckBuffer(L, 1);
+            dmBuffer::HBuffer hbuffer = UnpackLuaBuffer(buffer);
+
+            // get metadata entry name
+            dmhash_t entry_name = dmScript::CheckHashOrString(L, 2);
+
+            uint32_t count;
+            dmBuffer::ValueType valueType;
+            void* values;
+            dmBuffer::Result r = dmBuffer::GetMetaData(hbuffer, entry_name, &values, &count, &valueType);
+            if ( r != dmBuffer::RESULT_OK )
```

Comment:
I think I'd personally write this as to avoid nested ifs

```
                if ( r == dmBuffer::RESULT_METADATA_NOT_EXIST)
                {
                    lua_pushnil(L);  // nil for metadata entry
                    lua_pushnil(L); // nil for numbertype
                    return 2;
                }
                else if ( r != dmBuffer::RESULT_OK )
                {
                    return DM_LUA_ERROR("%s.%s: error getting metadata for buffer: %s", SCRIPT_LIB_NAME, SCRIPT_TYPE_NAME_BUFFERMETADATA, dmBuffer::GetResultString(r));
                }
```

But this is primarily a matter of taste I think :-)